vip-helpers/vip-utils.php

```diff
@@ -1283,17 +1283,17 @@ function wpcom_vip_can_use_shared_plugin( $plugin ) {
 /**
  * Helper function to check if we can load plugins or not.
  */
-function wpcom_vip_should_load_plugins() {
+function wpcom_vip_should_load_plugins( $_reset = false ) {
 	static $should_load_plugins;
 
-	if ( isset( $should_load_plugins ) ) {
+	if ( ! $_reset && isset( $should_load_plugins ) ) {
 		return $should_load_plugins;
 	}
 
 	$should_load_plugins = true;
 
 	// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
-	if ( wp_installing() && strtok( $_SERVER['REQUEST_URI'] ?? '', '?' ) !== '/wp-activate.php' ) {
+	if ( wp_installing() && ! str_ends_with( strtok( $_SERVER['REQUEST_URI'] ?? '', '?' ), '/wp-activate.php' ) ) {
```

Comment:
Why not using `basename()` which basically exists for cases like this and allows you to replace two function calls (`strtok` + ` str_ends_with`) with one?

Besides, can you guarantee that you'll never use this utility function before WordPress loads `compat.php` which defines `str_ends_with()` for PHP < 8.0? 

It would not be better to keep such utilities callable as early as possible and rely on "bare PHP" as much as possible?

For example, inside an `advanced-cache.php` dropin file, on PHP < 8.0, `str_ends_with` could not be defined, because `compat.php` is loaded _after_ that file [up to the pretty recent WP 6.2.2](https://github.com/WordPress/WordPress/blob/9e0f2faa28f1aa04a69e0b4eaa410a38adcd2e1e/wp-settings.php#L95-L107).

Moreover, the shim for that PHP 8.0 function was added in [WP 5.9](https://github.com/WordPress/WordPress/blob/90a29fbc4698e2959c7c35de9f4327fd98717bf7/wp-includes/compat.php#L474), are you absolutely sure you have no WP installation with a version older that that? I mean, we are talking here about a fatal error that can turn a producion site down.

Considering the alternative to avoid `str_ends_with()` is actually _less code_, and likely _more perfomant_ (because 1 VS 2 function calls), what's the reason of preferring it?

Reply:
@gmazzap I wish it were that easy. `REQUEST_URI` may contain a query string; `basename` will happily accept it. The original PR suggested to use [`parse_url()`](https://github.com/Automattic/vip-go-mu-plugins/pull/4861/files#diff-e3e0b5e7ec15b54034b133dc38c47759865fe8ef51de53b2cca48e81d2217cedR1295); it is much slower than `strtok()`.

Once we get rid of the query string, we have `basename()` + string comparison vs `str_ends_with()`

> Besides, can you guarantee that you'll never use this utility function before WordPress loads compat.php which defines str_ends_with() for PHP < 8.0?

Of course. We define `str_ends_with()` in `lib/helpers/php-compat.php`, which is loaded from `000-vip-init.php`. `lib/helpers/php-compat.php` is loaded before `vip-helpers/vip-utils.php`:

```php
// Polyfills
require_once __DIR__ . '/lib/helpers/php-compat.php';

// Load our development and environment helpers
require_once __DIR__ . '/vip-helpers/vip-notoptions-mitigation.php';
require_once __DIR__ . '/vip-helpers/vip-utils.php';
```

So when it comes to `wpcom_vip_should_load_plugins()`, we do have `str_ends_with()` defined.

> For example, inside an advanced-cache.php dropin file

`wpcom_vip_should_load_plugins()` won't be available to drop-ins because they are loaded prior to mu-plugins. And if/when `wpcom_vip_should_load_plugins()` is available, `str_ends_with()` is also available.

> Moreover, the shim for that PHP 8.0 function was added in [WP 5.9](https://github.com/WordPress/WordPress/blob/90a29fbc4698e2959c7c35de9f4327fd98717bf7/wp-includes/compat.php#L474), are you absolutely sure you have no WP installation with a version older that that?

Yes. WP 5.9 is the oldest supported version. Moreover, we define `str_ends_with()` even if WordPress does not.

> Considering the alternative to avoid str_ends_with() is actually less code, and likely more perfomant (because 1 VS 2 function calls), what's the reason of preferring it?

I would agree if it were `basename()` vs `strtok()` + `str_ends_with()`, but `basename()` alone unfortunately does not work. It needs either `strtok()` or `(wp_)parse_url()`.

And it turns out that `basename()` + string comparison is not faster than `str_ends_with()`: `basename()` has to scan for the directory separator while `str_ends_with()` already knows the length of the pattern. Then, you will have to compare the returned result with the pattern, and `str_ends_with()` already does that. For PHP 8+, this happens in the C code; this will be a bit faster than comparison in the userland.

And because this code runs on every request, we try to squeeze the tiniest bits of performance.

Reply:
Also, regarding

> wpcom_vip_should_load_plugins() won't be available to drop-ins because they are loaded prior to mu-plugins. 

Do I remember wrongly or the "lib" folder is loaded prior to mu-plugins as well? This is unrelevant for this topic because of your `php-compat.php`, more a curiosity on my end.

Reply:
> Do I remember wrongly or the "lib" folder is loaded prior to mu-plugins as well?

In `wp-config.php`:

```php
// Dependencies that should be loaded prior to VIP Config (custom error handler, Environment helpers, VIP_Request_Block)
if ( file_exists( __DIR__ . '/wp-content/mu-plugins/000-pre-vip-config/requires.php' ) ) {
        require_once __DIR__ . '/wp-content/mu-plugins/000-pre-vip-config/requires.php';
}

require( dirname( __FILE__ ) . '/config/wp-config.php' );
```

[`000-pre-vip-config/requires.php`](https://github.com/Automattic/vip-go-mu-plugins/blob/develop/000-pre-vip-config/requires.php) loads the following files:
```php
$files = [
	'/lib/wpcom-error-handler/wpcom-error-handler.php',
	'/lib/class-vip-request-block.php',
	'/lib/environment/class-environment.php',
	'/lib/helpers/environment.php',
	'/lib/utils/class-context.php',
];

$cli_files = [
	'/lib/helpers/wp-cli-db.php',
];
```

(all paths are relative to `wp-content/mu-plugins`).